libp2p/RLPXFrameWriter.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ This file is part of cpp-ethereum.
+ 
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/** @file RLPXFrameWriter.cpp
+ * @author Alex Leverington <nessence@gmail.com>
+ * @date 2015
+ */
+
+#include "RLPXFrameWriter.h"
+
+using namespace std;
+using namespace dev;
+using namespace dev::p2p;
+
+const uint16_t RLPXFrameWriter::EmptyFrameLength = h128::size * 3; // header + headerMAC + frameMAC
+const uint16_t RLPXFrameWriter::MinFrameDequeLength = h128::size * 4; // header + headerMAC + padded-block + frameMAC
+
+void RLPXFrameWriter::enque(RLPXPacket&& _p, PacketPriority _priority)
+{
+	if (!_p.isValid())
+		return;
+	WriterState& qs = _priority ? m_q.first : m_q.second;
+	DEV_GUARDED(qs.x)
+		qs.q.push_back(move(_p));
+}
+
+void RLPXFrameWriter::enque(uint8_t _packetType, RLPStream& _payload, PacketPriority _priority)
+{
+	enque(RLPXPacket(m_protocolType, (RLPStream() << _packetType), _payload), _priority);
+}
+
+size_t RLPXFrameWriter::mux(RLPXFrameCoder& _coder, unsigned _size, vector<bytes>& o_toWrite)
+{
+	static const size_t c_blockSize = h128::size;
+	static const size_t c_overhead = c_blockSize * 3; // header + headerMac + frameMAC
+	if (_size < c_overhead + c_blockSize)
+		return 0;
+
+	size_t ret = 0;
+	size_t frameLen = _size / 16 * 16;
+	bytes payload(0);
+	bool swapQueues = false;
+	while (frameLen >= c_overhead + c_blockSize)
+	{
+		bool highPending;
+		bool lowPending;
+		DEV_GUARDED(m_q.first.x)
+			highPending = !!m_q.first.q.size();
+		DEV_GUARDED(m_q.second.x)
+			lowPending = !!m_q.second.q.size();
+
+		if (!highPending && !lowPending)
+			return ret;
+		
+		// first run when !swapQueues, high > low, otherwise low > high
+		bool high = highPending && !swapQueues ? true : lowPending ? false : true;
```

Comment:
would be much better to write (!lowPending) instead of (lowPending ? false : true)